### ISSUE TITLE
When killing 1 node (out of 9 / 15 node cluster), Cluster ops drop dramatically for ~25 sec to ~30K

### ISSUE THREAD

**Issue by tomer-sandler** [scylladb/scylla#2756]
*Installation details*
Scylla version (or git commit hash): 2.0 RC3
Cluster size: 9 / 15
OS (RHEL/CentOS/Ubuntu/AWS AMI): AWS AMI

*Hardware details (for performance issues)*
Platform (physical/VM/cloud instance type/docker): VM i3.4XL
Disks: 2 x NVMe per VM, each 1900GB

I'm testing i3.4XL nodes and this problem happened  both in 9 node and 15 node clusters

*Scenario*
1. Using c-s default schema (RF=3 CL=one) populating 1TB data

Schema:
```
CREATE KEYSPACE "keyspace1" WITH replication = {
  'class': 'SimpleStrategy',
  'replication_factor': '3'
};
USE "keyspace1";
CREATE TABLE "standard1" (
  key blob,
  "C0" blob,
  "C1" blob,
  "C2" blob,
  "C3" blob,
  "C4" blob,
  PRIMARY KEY (key)
) WITH
  bloom_filter_fp_chance=0.010000 AND
  caching='KEYS_ONLY' AND
  comment='' AND
  dclocal_read_repair_chance=0.000000 AND
  gc_grace_seconds=864000 AND
  index_interval=128 AND
  read_repair_chance=0.100000 AND
  replicate_on_write='true' AND
  default_time_to_live=0 AND
  speculative_retry='99.0PERCENTILE' AND
  memtable_flush_period_in_ms=0 AND
  compaction={'class': 'SizeTieredCompactionStrategy'} AND
  compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'};
```
```
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_1-4.log -pop seq=1..340000000 &
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_2-4.log -pop seq=340000001..680000000 &
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_3-4.log -pop seq=680000001..1020000000 &
#nohup cassandra-stress write no-warmup n=340000000 cl=one -mode native cql3 -node [IPs] -rate threads=200 -log file=populate_1TB_4-4.log -pop seq=1020000001..1360000000 &
```

2. Waiting for compactions to end -> Ended.

3. Run 50% mixed load, CL=Q, from 8 c-s instanceds (4 VMs, 2 instances on each VM) to get steady 200K SLA
```
#nohup cassandra-stress mixed ratio\(write=1,read=1\) no-warmup duration=1h cl=quorum -pop dist=gaussian\(1..1360000000,680000000,227000000\) -mode native cql3 -node [IPs] -log file=50_mixed_nodefail_repair -rate threads=150 limit=25000/s &
```

4. After ~5 min, kill a NON-seed node
```
pkill -9 scylla
```
I notice cluster ops drop dramatically for ~25 sec to ~30K (see screenshot)

5. After ~5 min, I initiate repair on the node that failed
```
nodetool repair -pr keyspace1
```
On the 15 node cluster, single node repair took 17:20 min
```
[centos@ip-172-16-0-92 ~]$ nodetool repair -pr keyspace1
[2017-09-03 08:35:34,010] Starting repair command #1, repairing 1 ranges for keyspace keyspace1 (parallelism=SEQUENTIAL, full=true)

[2017-09-03 08:52:53,333] Repair session 1
[2017-09-03 08:52:53,338] Repair session 1 finished
```

15-node cluster:
![killnode_ops_drop](https://user-images.githubusercontent.com/25585995/30001751-c8160da0-909f-11e7-906f-73775eba0ade.PNG)

9-node cluster:
![9_node_cluster_killnode_ops_drop](https://user-images.githubusercontent.com/25585995/30001805-a6f4969e-90a1-11e7-96a2-a2244976487f.png)

**Comment by avikivity** [comment]
Most likely due to #2757.

**Comment by avikivity** [comment]
Try `with speculative_retry = 'ALWAYS'` or `with speculative_retry = '2ms'`.

**Comment by tomer-sandler** [comment]
@avikivity Avi, where do I add this in c-s client and Scylla? Is it a change in the Schema?

**Comment by tomer-sandler** [comment]
OK, I did ALTER TABLE and added speculative_retry = '2ms' and then repeated the test
This time there we no drop in latency both upon 'pkill -9 scylla', and upon VM Stop.

**Comment by tzach** [comment]
If speculative_retry is so good, should we make it the default?

**Comment by avikivity** [comment]
It is the default, just not implemented correctly.

**Comment by tomer-sandler** [comment]
@avikivity / @slivne - this bug has been open for 8.5 months (found while using 2.0 RC3). Do you know if this is fixed for 2.1.3 and we can use the default C-S schema, or the recommended change by Avi is still needed?
Reason I'm asking is that we are about to conduct new benchmark test for Scylla 2.1.3 vs. C* 3.11. It will also include killing a node, bringing it back and running repair.

**Comment by tomer-sandler** [comment]
@avikivity / @slivne / @tzach  ignore my last comment, I see that the related issue #2757, is closed. So we probably need to close this bug as well.

**Comment by slivne** [comment]
yes we can close it
 
please note that we have a different issue about speculative retry that is still open #3156 - but its not related to killing a node

**Comment by tomer-sandler** [comment]
@avikivity / @slivne / @tzach 
I just reproduced this issue in Scylla 2.1.3.

6-node cluster (i3.8XL)
200K IOps, killing 1 node and I see ~75% IOps drop for ~30. Looks like exactly the same problematic behaviour.

Same Schema (with ```'99.0PERCENTILE'```)
```
CREATE KEYSPACE "keyspace1" WITH replication = {
  'class': 'SimpleStrategy',
  'replication_factor': '3'
};
USE "keyspace1";
CREATE TABLE "standard1" (
  key blob,
  "C0" blob,
  "C1" blob,
  "C2" blob,
  "C3" blob,
  "C4" blob,
  PRIMARY KEY (key)
) WITH
  bloom_filter_fp_chance=0.010000 AND
  caching='KEYS_ONLY' AND
  comment='' AND
  dclocal_read_repair_chance=0.000000 AND
  gc_grace_seconds=864000 AND
  index_interval=128 AND
  read_repair_chance=0.100000 AND
  replicate_on_write='true' AND
  default_time_to_live=0 AND
  speculative_retry='99.0PERCENTILE' AND
  memtable_flush_period_in_ms=0 AND
  compaction={'class': 'SizeTieredCompactionStrategy'} AND
  compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'};
```

**Comment by slivne** [comment]
@tomer-sandler - to know if its the exact same thing or not - can you try to change the speculative_retry like you have done before

@gleb-cloudius any inputs ?

**Comment by tomer-sandler** [comment]
@slivne, once repair completes I'll ALTER the schema and repeat the test with ```specilative_retry ='2ms'```
@gleb-cloudius, as for the 99th percentile -> see screenshot

![image](https://user-images.githubusercontent.com/25585995/40055270-03100ea6-5850-11e8-9f2b-30dcbde9e177.png)

**Comment by gleb-cloudius** [comment]
Hard to see the value exactly. Can you post it here (the value before
the jump). Also do you have max latency graph?

--
			Gleb.

**Comment by tomer-sandler** [comment]
Background workload: 50% Writes (updates) + 50% Reads.

Write 99th: ~7 ms

![image](https://user-images.githubusercontent.com/25585995/40056620-a29c96e8-5854-11e8-9f96-1ff35e7c9a16.png)


Read 99th: ~13ms

![image](https://user-images.githubusercontent.com/25585995/40056676-bf46a7ac-5854-11e8-9f6d-59143ba0f530.png)


I Duplicated the matric and modified from 0.99 to 1.0, to get the MAX but every time I refresh I see different numbers for the same spot in time, so I can't tell you for sure. Sometimes it's 33-66 ms and sometimes I see also 1 node with ~250ms and another node on ~8 sec.

**Comment by gleb-cloudius** [comment]
What is the max read latency when 99 picks at 400ms?

Try 13ms instead of 2ms. If it will work it may mean that we decay
99 percentile to fast.

--
			Gleb.

**Comment by tomer-sandler** [comment]
You originally asked for the 99th before the node was killed (11:44:15) and these are the results I posted.
When the 99th jumps to 400ms (11:44:30) that already when the node is dead.

In order to see the MAX I did the following:
1. Edit the original metrics:
```histogram_quantile(0.99, sum(rate(scylla_storage_proxy_coordinator_write_latency_bucket{instance=~"[[node]]", shard=~"[[shard]]"}[30s])) by ([[by]], le))```
```histogram_quantile(0.99, sum(rate(scylla_storage_proxy_coordinator_read_latency_bucket{instance=~"[[node]]", shard=~"[[shard]]"}[30s])) by ([[by]], le))```
2. Change the ```0.99``` to ```1.0```

I don't know the MAX from the Grafana as every refresh it displays different results.
Maybe I'm doing it wrong. Do you know the correct metric syntax?

**Comment by gleb-cloudius** [comment]
On Tue, May 15, 2018 at 06:25:48AM -0700, tomer-sandler wrote:
> You originally asked for the 99th before the node was killed (11:44:15) and these are the results I posted.
> When the 99th jumps to 400ms (11:44:30) that already when the node is dead.
> 
I am interested in latencies of other nodes, not the one that got
killed.

> In order to see the MAX I did the following:
> 1. Edit the original metrics:
> ```histogram_quantile(0.99, sum(rate(scylla_storage_proxy_coordinator_write_latency_bucket{instance=~"[[node]]", shard=~"[[shard]]"}[30s])) by ([[by]], le))```
> ```histogram_quantile(0.99, sum(rate(scylla_storage_proxy_coordinator_read_latency_bucket{instance=~"[[node]]", shard=~"[[shard]]"}[30s])) by ([[by]], le))```
> 2. Change the ```0.99``` to ```1.0```
> 
> I don't know the MAX from the Grafana as every refresh it displays different results.
> Maybe I'm doing it wrong. Do you know the correct metric syntax? 
> 
Never used histogram in grafana. May be @glauber knows.

--
			Gleb.

**Comment by tomer-sandler** [comment]
The results I posted are for all the nodes, so you can see that just before it was killed the 99th AVG (avg means the results from all 6 nodes) was as I wrote before.
Write 99th: ~7 ms
Read 99th: ~13ms

I was able to find the spot where the latency spikes in the C-S output log - you can see the spike to ~15K-30K ms (1.5-3 sec) a bit before that it's ~200-300 ms, and a little bit more before that it's ~20-40 ms

1st 4 c-s clients:
![image](https://user-images.githubusercontent.com/25585995/40060635-c2723e4a-585f-11e8-95e4-c98ff3ecd042.png)

2nd 4 c-s clients:
![image](https://user-images.githubusercontent.com/25585995/40060870-3a815402-5860-11e8-8fe0-452bbc87ac17.png)

**Comment by tomer-sandler** [comment]
Ran the test again with ```speculative_retry ='13ms'``` (per Gleb's request).  Same background workload (50% read  50% write, total 200K IOps). Still got 50% drop in throughput for ~30 sec. Gleb and I also noticed there are 40 read timeout per second during that time, and that writes are also affected, not only reads.

I repeated the test with ```speculative_retry ='2ms'``` and I got the same outcome as with ```'13ms'```. 50% throughput drop. this time for ~45 sec.

Might be different issue...
